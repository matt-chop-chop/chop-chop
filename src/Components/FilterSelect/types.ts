type SelectOption = {
  label: string;
  value: string;
  description?: string;
};
