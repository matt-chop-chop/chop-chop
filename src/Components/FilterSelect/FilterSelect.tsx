import { styles } from "./style";
import React from "react";
import { InfoOutlineIcon } from "@chakra-ui/icons";
import { Flex, Tooltip } from "@chakra-ui/react";
import Select, { components, OptionProps } from "react-select";

const Option = ({ ...props }: OptionProps<SelectOption, false>) => {
  const { description, label } = props.data;

  return (
    <components.Option {...props}>
      <Flex alignItems="center" justifyContent="space-between">
        {label}
        {description && (
          <Tooltip hasArrow label={description} placement="left">
            <InfoOutlineIcon mr={3} />
          </Tooltip>
        )}
      </Flex>
    </components.Option>
  );
};

export type SelectOption = {
  label: string;
  value: string;
  description?: string;
};

type SelectItem = {
  name: string;
  description?: string;
};

export type FilterSelectProps = {
  options: SelectItem[];
  onSelectedChange: (selectedItem: string) => void;
  id?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
};

const FilterSelect = ({
  options,
  onSelectedChange,
  id = "select",
  isDisabled = false,
  isLoading = false,
}: FilterSelectProps) => {
  const getFormattedOptions = (options: SelectItem[]): SelectOption[] => {
    return options.map((option) => {
      return {
        label: option.name,
        value: option.name,
        description: option?.description,
      };
    });
  };

  const handleOnChange = (newValue: SelectOption | null) => {
    if (newValue !== null) onSelectedChange(newValue.value);
  };

  return (
    <Select<SelectOption, false>
      components={{
        DropdownIndicator: () => null,
        IndicatorSeparator: () => null,
        Option,
      }}
      id={id}
      instanceId={id}
      isDisabled={isDisabled}
      isLoading={isLoading}
      onChange={handleOnChange}
      options={getFormattedOptions(options)}
      placeholder="Select a value..."
      styles={styles}
    />
  );
};

export default FilterSelect;
