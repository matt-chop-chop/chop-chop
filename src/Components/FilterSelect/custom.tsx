import { InfoOutlineIcon } from "@chakra-ui/icons";
import { Flex, Tooltip } from "@chakra-ui/react";
import { StylesConfig, components, OptionProps } from "react-select";

export const Option = ({ ...props }: OptionProps<SelectOption, false>) => {
  const { description, label } = props.data;

  return (
    <components.Option {...props}>
      <Flex alignItems="center" justifyContent="space-between">
        {label}
        {description && (
          <Tooltip hasArrow label={description} placement="left">
            <InfoOutlineIcon mr={3} />
          </Tooltip>
        )}
      </Flex>
    </components.Option>
  );
};

export const styles: StylesConfig<SelectOption, false> = {
  control: (_, { isDisabled, isFocused }) => {
    const borderColor = isFocused ? "var(--chakra-colors-light-emphasis)" : "";
    const opacity = isDisabled ? 0.4 : 1;

    return {
      alignItems: "center",
      background: "var(--chakra-colors-light-background)",
      border: "2px solid var(--chakra-colors-light-text)",
      borderColor,
      borderRadius: "4px",
      color: "var(--chakra-colors-light-text)",
      cursor: "pointer",
      display: "flex",
      fontSize: "16px",
      fontWeight: 400,
      lineHeight: "28px",
      opacity,
    };
  },
  menu: (provided) => ({
    ...provided,
    margin: 0,
  }),
  menuList: (provided) => ({
    ...provided,
    background: "var(--chakra-colors-light-text)",
    borderRadius: "none",
    padding: 0,
  }),
  option: (provided, { isFocused }) => ({
    ...provided,
    background: isFocused
      ? "var(--chakra-colors-light-emphasis)"
      : "var(--chakra-colors-light-text)",
    color: "var(--chakra-colors-light-background)",
    cursor: "pointer",
    fontSize: "16px",
    fontWeight: 400,
    lineHeight: "28px",

    ":active": {
      backgroundColor: "var(--chakra-colors-light-emphasis)",
    },

    ":hover": {
      background: "var(--chakra-colors-light-emphasis)",
    },
  }),
  valueContainer: (provided) => ({
    ...provided,
    paddingX: "14px",
    paddingY: "4px",
  }),
};
