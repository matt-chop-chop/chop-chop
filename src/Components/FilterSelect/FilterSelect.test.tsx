import { shallow } from "enzyme";
import React from "react";
import FilterSelect, { FilterSelectProps, SelectOption } from "./FilterSelect";
import { ActionMeta, SingleValue } from "react-select";

const onSelectChange = jest.fn();

const props: FilterSelectProps = {
  options: [
    { name: "Test 1" },
    { name: "Test 2", description: "This is a test" },
  ],
  onSelectedChange: onSelectChange,
};

describe("FilterSelect", () => {
  it("is exported from FilterSelect", () => {
    expect(FilterSelect).toBeDefined();
  });

  it("renders as expected ", () => {
    const wrapper = shallow(<FilterSelect {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("calls onSelectedChange when onChange is triggered ", () => {
    const wrapper = shallow(<FilterSelect {...props} />);

    const Select = wrapper.find("ForwardRef");
    expect(Select).toHaveLength(1);

    Select.simulate("change", { value: "Test 1" });

    expect(onSelectChange).toHaveBeenCalled();
  });
});
