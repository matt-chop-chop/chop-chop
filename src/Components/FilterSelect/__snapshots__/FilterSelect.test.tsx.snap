// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSelect renders as expected  1`] = `
<ForwardRef
  components={
    {
      "DropdownIndicator": [Function],
      "IndicatorSeparator": [Function],
      "Option": [Function],
    }
  }
  id="select"
  instanceId="select"
  isDisabled={false}
  isLoading={false}
  onChange={[Function]}
  options={
    [
      {
        "description": undefined,
        "label": "Test 1",
        "value": "Test 1",
      },
      {
        "description": "This is a test",
        "label": "Test 2",
        "value": "Test 2",
      },
    ]
  }
  placeholder="Select a value..."
  styles={
    {
      "control": [Function],
      "menu": [Function],
      "menuList": [Function],
      "option": [Function],
      "valueContainer": [Function],
    }
  }
/>
`;
